internal/clients/stubs/azure_stub.go

```diff
@@ -41,21 +44,64 @@ func (stub *AzureClientStub) Status(ctx context.Context) error {
 	return nil
 }
 
-func (stub *AzureClientStub) CreateVM(ctx context.Context, location string, resourceGroupName string, imageID string, pubkey *models.Pubkey, instanceType clients.InstanceTypeName, vmName string, userData []byte) (*string, error) {
+func (stub *AzureClientStub) CreateVM(ctx context.Context, vmParams clients.AzureInstanceParams, vmName string) (*string, error) {
 	id := strconv.Itoa(len(stub.createdVms) + 1)
 
 	vm := armcompute.VirtualMachine{
 		ID:       &id,
 		Name:     &vmName,
-		Location: &location,
-		Properties: &armcompute.VirtualMachineProperties{
-			UserData: ptr.To(string(userData)),
-		},
+		Location: &vmParams.Location,
 	}
 	stub.createdVms = append(stub.createdVms, &vm)
 	return &id, nil
 }
 
+func (stub *AzureClientStub) CreateVMs(ctx context.Context, vmParams clients.AzureInstanceParams, amount int64, vmNamePrefix string) ([]*string, error) {
+	vmIds := make([]*string, amount)
+	resumeTokens := make([]string, amount)
+	var i int64
+	var err error
+	for i = 0; i < amount; i++ {
+		vmName := fmt.Sprintf("%s-%d", vmNamePrefix, int64(len(stub.startedVms))+i)
+		resumeTokens[i], err = stub.BeginCreateVM(ctx, vmParams, vmName)
+		if err != nil {
+			return vmIds, err
+		}
+	}
+	for i = 0; i < amount; i++ {
+		vmIds[i], err = stub.WaitForVM(ctx, resumeTokens[i])
+		if err != nil {
+			return vmIds, err
+		}
+	}
+
+	return vmIds, nil
+}
```

Comment:
I think since you dropped exporting begin/wait functions, you can also simplify the stub  :-D

If you drop the other function and only keep CreateVMs then you can further simplify it.

Also, you do not have to really begin/wait in a stub, stub can simply return "yep, all were launched" :-)

Reply:
I kept the stub ready for what I anticipate will happen - exposing the begin function and don't wait for the machines to come up.
We should discuss unifying this, as for AWS we just get ack from AWS "we are working on it" and we are not waiting for the fullfilment and that's IMHO the right approach, so I'd discuss it first and simplify the stub once we are sure we won't want only the "BeginCreate" :)
But if you don't like it I can save it in local branch ofc xD I just thought why not mimic the real thing, when it's already done :)

Reply:
I don’t see us not waiting for VM to launch, these are very cheap jobs basically polling HTTP endpoint, workers can handle thousands of them at one time. I don’t see this as a problem. We might update our job to send a notification or something and update our UI but I think an API with just "CreateVMs" is okay.

Reply:
I was thinking exactly of sending memo to UI and I think that easiest way to do this is to split it in the client API and be able to figure out in the job where we are at and that can be done (cleanly) only if we surface the split of planning the launch and waiting for it ;)